src/main/java/net/snowflake/ingest/streaming/internal/SnowflakeStreamingIngestChannelInternal.java

```diff
@@ -128,41 +105,24 @@
       OpenChannelRequest.OnErrorOption onErrorOption,
       Constants.BdecVersion bdecVersion,
       BufferAllocator allocator) {
-    this.channelName = name;
-    this.dbName = dbName;
-    this.schemaName = schemaName;
-    this.tableName = tableName;
-    this.offsetToken = offsetToken;
-    this.channelSequencer = channelSequencer;
-    this.rowSequencer = new AtomicLong(rowSequencer);
-    this.isValid = true;
     this.isClosed = false;
     this.owningClient = client;
     this.allocator = allocator;
-    this.rowBuffer = createRowBuffer(bdecVersion);
-    this.encryptionKey = encryptionKey;
-    this.encryptionKeyId = encryptionKeyId;
-    this.onErrorOption = onErrorOption;
-    logger.logInfo("Channel={} created for table={}", this.channelName, this.tableName);
-  }
-
-  private RowBuffer<T> createRowBuffer(Constants.BdecVersion bdecVersion) {
-    // TODO: The circular dependency SnowflakeStreamingIngestChannelInternal <-> RowBuffer
-    // (SNOW-657667)
-    // can be probably reconsidered
-    switch (bdecVersion) {
-      case ONE:
-        //noinspection unchecked
-        return (RowBuffer<T>)
-            new ArrowRowBuffer((SnowflakeStreamingIngestChannelInternal<VectorSchemaRoot>) this);
-      case THREE:
-        //noinspection unchecked
-        return (RowBuffer<T>)
-            new ParquetRowBuffer((SnowflakeStreamingIngestChannelInternal<ParquetChunkData>) this);
-      default:
-        throw new SFException(
-            ErrorCode.INTERNAL_ERROR, "Unsupported BDEC format version: " + bdecVersion);
-    }
+    this.channelContext =
+        new ChannelContext(
+            name, dbName, schemaName, tableName, channelSequencer, encryptionKey, encryptionKeyId);
+    AbstractRowBuffer.BufferConfig bufferConfig =
```

Comment:
I feel like the channelContext and bufferConfig are confusing, for example, the onErrorOption, offsetToken or rowSequencer should be associated with a channel instead of a configuration on a buffer, should we just create a channelContext and passing it around?

Reply:
From what I see, onErrorOption, offsetToken or rowSequencer are basically used and managed mostly in the buffer.

> should we just create a channelContext and passing it around?

they have pretty different contents, I do not like mixing concerns into one object that basically results in couplings that this PR tries to address. I think each class should contain only what it needs to serve its purpose.

`BufferingContext` (renamed `BufferConfig`) is about setting up the buffer.
`ChannelContext` (I renamed it to ChannelAttributes) is to describe the channel (identification, encryption), it does not have much to do with buffering directly.

Reply:
Thanks for the reply. I do understand that the fields like `onErrorOption`, `offsetToken` or `rowSequencer` is just used/updated in the buffer, but these fields should still be associated with a channel and I find it confusing if we associate it with the buffer instead of channel. It's not clear to the readers that channel and buffer are 1:1 mapped, what if we update the logic in the future that all the channels belong to the same table are inserting into the same buffer? If there is no better way, I would prefer to keep this logic as is since it's easier to understand and live with the circular dependency. WDYT?

Reply:
Shared object sounds good to me, thanks!

Reply:
As agreed with @sfc-gh-azagrebin, I pushed a commit that does that. PTAL.